### PR TITLE
AA-453: Support EIP-7702 tuples in ERC-4337 bundlers

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: packages/bundler/src/BundlerServer.ts [hunk 1]
```diff
@@ -277,9 +288,15 @@ export class BundlerServer {
         result = await this.methodHandler.getSupportedEntryPoints()
         break
       case 'eth_sendUserOperation':
-        result = await this.methodHandler.sendUserOperation(params[0], params[1])
+        if (!this.config.eip7702Support && params[2] != null) {
+          throw new Error('EIP-7702 tuples are not supported')
+        }
+        result = await this.methodHandler.sendUserOperation(params[0], params[1], params[2])
```
[the remark this answers is not in the document]
Comment (drortirosh) [reply packages/bundler/src/BundlerServer.ts:294]: (just like it is an optional field in AATransaction of 7560)

[the remark this answers is not in the document]
Comment (forshtat) [reply packages/bundler/src/BundlerServer.ts]: Fixed.

File: packages/bundler/src/modules/BundleManager.ts [hunk 2]
```diff
@@ -98,12 +103,13 @@ export class BundleManager implements IBundleManager {
    * after submitting the bundle, remove all UserOps from the mempool
    * @return SendBundleReturn the transaction and UserOp hashes on successful transaction, or null on failed transaction
    */
-  async sendBundle (userOps: UserOperation[], beneficiary: string, storageMap: StorageMap): Promise<SendBundleReturn | undefined> {
+  async sendBundle (userOps: UserOperation[], eip7702Tuples: EIP7702Tuple[], beneficiary: string, storageMap: StorageMap): Promise<SendBundleReturn | undefined> {
     try {
       const feeData = await this.provider.getFeeData()
       // TODO: estimate is not enough. should trace with validation rules, to prevent on-chain revert.
+      const type = eip7702Tuples.length > 0 ? TX_TYPE_EIP_7702 : TX_TYPE_EIP_1559
       const tx = await this.entryPoint.populateTransaction.handleOps(userOps.map(packUserOp), beneficiary, {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/bundler/src/modules/BundleManager.ts:115]: Please see the code now. You were looking at a partially checked in code.

File: packages/bundler/src/modules/BundleManager.ts [hunk 3]
```diff
@@ -29,11 +31,19 @@ import { IBundleManager } from './IBundleManager'
...
+import { EOACode7702Transaction } from '@ethereumjs/tx'
+import { AuthorizationList, EOACode7702TxData } from '@ethereumjs/tx/src/types'
+import { PrefixedHexString } from '@ethereumjs/util'
+import { toRlpHex } from '@account-abstraction/utils/dist/src/interfaces/EIP7702Authorization'
 
 const debug = Debug('aa.exec.cron')
 
 const THROTTLED_ENTITY_BUNDLE_COUNT = 4
 
+const TX_TYPE_EIP_7702 = 4
```
Comment (shahafn) [inline packages/bundler/src/modules/BundleManager.ts:44]: Should it be in config or something? We already know that this type means different things on different networks

File: packages/bundler/src/modules/BundleManager.ts [hunk 4]
```diff
@@ -168,6 +184,67 @@ export class BundleManager implements IBundleManager {
...
+      data: tx.data!.toString() as PrefixedHexString,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      chainId: `0x${tx.chainId!.toString(16)}`,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      maxPriorityFeePerGas: tx.maxPriorityFeePerGas!.toHexString() as PrefixedHexString,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      maxFeePerGas: tx.maxPriorityFeePerGas!.toHexString() as PrefixedHexString,
+      accessList: [],
+      authorizationList
+    }
```
Comment (shahafn) [inline packages/bundler/src/modules/BundleManager.ts:232]: This is an issue when we have unused gas penalty

File: packages/bundler/src/modules/BundleManager.ts [hunk 5]
```diff
@@ -330,13 +410,27 @@ export class BundleManager implements IBundleManager {
       }
       mergeStorageMap(storageMap, validationResult.storageMap)
 
-      const newBundleGas = entry.userOpMaxGas.add(bundleGas)
-      bundleGas = newBundleGas
+      for (const eip7702Authorization of entry.userOp.authorizationList ?? []) {
```
Comment (shahafn) [inline packages/bundler/src/modules/BundleManager.ts:413]: Please extract to a function. This function is already too big to read

File: packages/bundler/src/modules/BundleManager.ts (hunk 3 above)
Comment (forshtat) [reply packages/bundler/src/modules/BundleManager.ts:44]: I don't think so, the EIP-7702 once merged into mainnet will be a part of the EVM and any chain that wants Tx 4 to mean anything else will have to switch. I don't know if there is any reasonable chain that will not be compatible, do you?

File: packages/bundler/src/modules/BundleManager.ts [hunk 6]
```diff
@@ -168,6 +184,67 @@ export class BundleManager implements IBundleManager {
     }
   }
 
+  // TODO: this is a temporary patch until ethers.js adds EIP-7702 support
```
[the remark this answers is not in the document]
Comment (forshtat) [reply packages/bundler/src/modules/BundleManager.ts:187]: Yes, I would prefer to use Ethers.js and not import Ethereumjs as well, but for the server the difference is insignificant (in the browser you wouldn't want two ether libraries) and Ethers.js is not ready with 7702 yet.

**Review by shahafn — APPROVED** [review]

**Pull request merged**